security/rbac/engine/engine.go

```diff
@@ -0,0 +1,195 @@
+}
+
+// AuthorizationDecision is the output of CEL-based authorization engines.
+// If decision is allow or deny, policyNames will either contain the names of
+// all the policies matched in the engine that permitted the action, or be
+// empty as the decision was made after all conditions evaluated to false.
+// If decision is unknown, policyNames will contain the list of policies that
+// evaluated to unknown.
+type AuthorizationDecision struct {
+	decision    Decision
+	policyNames []string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	// TODO(@ezou): implement the conversion from expr to CEL program.
+	return nil
+}
+
+// Evaluates a CEL expression. Returns true if CEL evaluation returns true. Returns false if
+// CEL evaluation returns false. Returns an error if CEL evaluation returns unknown or an
+// unsuccessful evaluation.
+func evaluateCondition(condition *cel.Program, evalMap map[string]interface{}) (bool, error) {
+	// TODO(@ezou): implement the matching logic using CEL library.
+	return false, nil
+}
+
+// celEngine is the struct for an engine created from one RBAC proto.
+type celEngine struct {
+	action     pb.RBAC_Action
+	conditions map[string]*cel.Program
+}
+
+// Creates a new celEngine from an RBAC policy proto.
+func newCelEngine(rbac pb.RBAC) celEngine {
+	action := rbac.Action
+	conditions := make(map[string]*cel.Program)
+	for policyName, policy := range rbac.Policies {
+		conditions[policyName] = exprToProgram(policy.Condition)
+	}
+	return celEngine{action, conditions}
+}
+
+// Returns the decision of an engine based on whether or not AuthorizationArgs is a match,
+// i.e. if engine's action is ALLOW and match is true, we will return DecisionAllow;
+// if engine's action is ALLOW and match is false, we will return DecisionDeny.
+func getDecision(engine celEngine, match bool) Decision {
+	if engine.action == pb.RBAC_ALLOW && match || engine.action == pb.RBAC_DENY && !match {
+		return DecisionAllow
+	}
+	return DecisionDeny
+}
+
+// Returns the authorization decision of a single engine based on evalMap.
+// If any policy matches, the decision matches the engine's action, and the list of
+//  matching policy names will be returned.
+// Else if any policy is missing attributes, the decision is unknown, and the list of
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, the decision is the opposite of the engine's action, i.e. an ALLOW engine
+//  will return DecisionDeny, and vice versa.
+func (engine celEngine) evaluate(evalMap map[string]interface{}) (Decision, []string) {
+	matchingPolicyNames := []string{}
+	unknownPolicyNames := []string{}
+	for policyName, condition := range engine.conditions {
```

Comment:
@TristonianJones In IAM policy verification with multiple CEL expressions. Once there is a match, do we short circuit it or continue evaluating all of them so that we can get a full list of matching expressions?

Reply:
You would short-circuit. The only trick here is that you'd continue evaluating if the condition is an `error` or `types.Unknown` (if you're using partial state) until you got to a `true` or had exhausted the condition set.

Reply:
That aligns with my expectation. Thanks Tristan for confirming.